src/api.rs

```diff
@@ -0,0 +1,119 @@
+};
+
+pub type Pipeline = compositor_pipeline::Pipeline<RtpReceiver, RtpSender>;
+
+#[derive(Serialize, Deserialize)]
+pub struct RegisterInputRequest {
+    pub id: InputId,
+    pub port: u16,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RegisterOutputRequest {
+    pub id: OutputId,
+    pub port: u16,
+    pub ip: String,
+    pub resolution: Resolution,
+    pub encoder_settings: EncoderSettings,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum Request {
+    RegisterInput(RegisterInputRequest),
+    RegisterOutput(RegisterOutputRequest),
+    UnregisterInput {
+        id: InputId,
+    },
+    UnregisterOutput {
+        id: OutputId,
+    },
+    UpdateScene(Arc<SceneSpec>),
+    RegisterTransformation {
+        key: TransformationRegistryKey,
+        transform: TransformationSpec,
+    },
+    Init(pipeline::Options),
+    Start,
+}
+
+pub struct Api {
+    pipeline: Pipeline,
+}
+
+impl Api {
+    pub fn new(opts: pipeline::Options) -> Result<Api> {
+        Ok(Api {
+            pipeline: Pipeline::new(opts)?,
+        })
+    }
+
+    pub fn handle_request(&mut self, request: Request) -> Result<()> {
+        match request {
+            Request::Init(_) => Err(anyhow!("Video compositor is already initialized.")),
+            Request::RegisterInput(request) => self.register_input(request),
+            Request::UnregisterInput { id } => Ok(self.pipeline.unregister_input(&id)?),
+            Request::RegisterOutput(request) => self.register_output(request),
+            Request::UnregisterOutput { id } => Ok(self.pipeline.unregister_output(&id)?),
+            Request::Start => {
+                self.pipeline.start();
+                Ok(())
+            }
+            Request::UpdateScene(scene_spec) => Ok(self.pipeline.update_scene(scene_spec)?),
+            Request::RegisterTransformation {
+                key,
+                transform: spec,
+            } => Ok(self.pipeline.register_transformation(key, spec)?),
+        }
+    }
+
+    fn register_output(&mut self, request: RegisterOutputRequest) -> Result<()> {
+        let RegisterOutputRequest {
+            id,
+            port,
+            resolution,
+            encoder_settings,
+            ip,
+        } = request;
+
+        self.pipeline.register_output(
+            id,
+            rtp_sender::Options {
+                port,
+                ip,
+                resolution,
+                encoder_settings,
+            },
+        )?;
+
+        Ok(())
+    }
+
+    fn register_input(&mut self, request: RegisterInputRequest) -> Result<()> {
+        let RegisterInputRequest { id, port } = request;
+
+        //if let Some((node_id, _)) = state.inputs.iter().find(|(_, input)| input.port == port) {
+        //    return Err(anyhow!(
+        //        "Failed to register input with {id}. Port {port} is already use by node {node_id}"
+        //    ));
+        //}
```

Comment:
Should we keep this? `pipeline.register_input()` should be responsible for that

Reply:
I'll address that in follow-up, I need to think about it. Problem here is that pipeline can't be aware of ports